### PR TITLE
Set chunked Transfer-Encoding when using `send`

### PR DESCRIPTION
**Pull request by Deluvi**
Related issue: #43 

Previously, using `.send()` on `Request` would require to set either the Transfer-Encoding or the Content-Length header.

In an effort to provide better ergonomics for this library and to avoid making users fall in a not-so obvious pitfall, the library should build a valid Request without asking the user to mess around with the headers.

This PR attempts to fix this issue: if the user use `.send()` to provide an unknown sized reader, the chunked Transfer-Encoding will be used. Of course, there are prior checks to ensure we do not override the user wish, like if the user already set a Content-Length or a Transfer-Encoding.

This PR also fix an edge case where the Content-Length would not be automatically set if the Transfer-Encoding is set but not chunked.

### CODE REVIEW

**Comment by Deluvi** [comment]
I have three remarks about this PR:

The Content-Length header is not set if the size of the body is 0 to not break the previous behavior. The HTTP standard does ask for a Content-Length header in some HTTP methods requests such as POST, PUT and PATCH. This might be better addressed on another PR.

[According to some RFC](https://tools.ietf.org/html/rfc7230#section-3.3.1):
> A sender MUST NOT send a Content-Length header field in any message that contains a Transfer-Encoding header field.

It does not make sense in the case where the Transfer-Encoding is something else than `chunked`. [According to the Mozilla wiki](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#Directives), this restriction seems to only apply on the `chunked` encoding. On this PR, I took Mozilla side and automatically set the Content-Length if the size is known and if the Transfer-Encoding is not `chunked`.

Finally, I do not check if the user has set both the `chunked` Transfer-Encoding and the Content-Length headers, which is not allowed in the HTTP1.1 standard.

File: src/body.rs [hunk 1]
```diff
@@ -65,7 +65,7 @@ impl SizedReader {
 impl Payload {
     pub fn into_read(self) -> SizedReader {
         match self {
-            Payload::Empty => SizedReader::new(None, Box::new(empty())),
+            Payload::Empty => SizedReader::new(Some(0), Box::new(empty())),
```
Comment (algesten) [inline src/body.rs:68]: Is this change needed for the overall logic of this PR?

Comment (Deluvi) [reply src/body.rs:68]: I needed a way to differentiate `Payload::Reader` and `Payload::Empty`, `None` allowed me to express that the `SizedReader` has an unknown size, as opposed to `Some(0)` which express an empty Reader.

**Comment by algesten** [comment]
I've been thinking about this. There are basically three ways the user can provide a body:

1. A call without a body, GET, HEAD, etc.
2. Send a body with unknown size (results in chunked)
3. Send a body with known size (sets content-length)

The code did already not capture this intention because it is using an `Option<usize>` of value `None` covered both case 1 and 2. Changing case 1 to be represented as `Some(0)` feels wrong – it means we can't distinguish that case from say sending a 0 byte file. It's no worse than today's code, but it's a sideways move.

One intention of this PR, and a potential follow-up PR is to bring us closer to spec. In that spirit, regardless of how that next PR will look, I think we should introduce an enum that preserves the user intent so we can properly deal with the logic later.

Here's a suggestion, totally up for bike shedding:

```rust
pub enum BodySize {
  NoBody,
  Unknown,
  Known(u64),
}
```

I deliberately chose `u64` over `usize`.

**Comment by jsha** [comment]
@Deluvi said:
> > A sender MUST NOT send a Content-Length header field in any message that contains a Transfer-Encoding header field.
> 
> It does not make sense in the case where the Transfer-Encoding is something else than chunked. According to the Mozilla wiki, this restriction seems to only apply on the chunked encoding. On this PR, I took Mozilla side and automatically set the Content-Length if the size is known and if the Transfer-Encoding is not chunked.

A little earlier [in that same section](https://tools.ietf.org/html/rfc7230#section-3.3.1), RFC 7230 says:

> If any transfer coding
   other than chunked is applied to a request payload body, the sender
   MUST apply chunked as the final transfer coding to ensure that the
   message is properly framed. 

So it does make sense to the Content-Length MUST NOT coexist with Transfer-Encoding, because if there is any Transfer-Encoding header, the last element of the last Transfer-Encoding header MUST be "chunked."

I think the right behavior for `.send()` here is:

 - If "Transfer-Encoding" and "Content-Length" were both set by the user, error.
 - If "Content-Length" was set by the user, copy the bytes without using chunked encoding (this is current behavior; I documented it wrong in #73).
 - If "Transfer-Encoding" was set by the user, and the last element of the last Transfer-Encoding header was "chunked," use chunked encoding and don't add any headers.
 - Otherwise, use chunked encoding and add a "Transfer-Encoding: chunked" header.

@algesten said:
>     pub enum BodySize {
>       NoBody,
>       Unknown,
>       Known(u64),
>     }
> I deliberately chose u64 over usize.

This seems like a good idea. Out of curiosity, why u64 over usize?

Also, to make it explicit since it took me a minute to figure out: the reason to make this distinction is so that on NoBody requests (GET etc) we want to send neither Content-Encoding nor Transfer-Encoding.

**Comment by algesten** [comment]
@jsha 

> This seems like a good idea. Out of curiosity, why u64 over usize?

Since usize is either 32 or 64 bits depending on arch, it makes a lot of sense that for example `Vec::len()` returns an usize. It's highly unlikely that this would ever be a problem, but the size of the body would be arbitrarily limited to 4GB if ureq is compiled on a 32 bit arch.

> the reason to make this distinction is because on NoBody requests (GET etc) we want to send neither Content-Encoding nor Transfer-Encoding.

Exactly.

**Comment by Deluvi** [comment]
Thank you you two for your feedback, very insightful !

I was on the edge to write something like `BodySize` when I was writing this PR. I tried to keep it simple so I didn't add it, but since you are suggesting it, I will add it gladly ! :smile: 

> A little earlier in that same section, RFC 7230 says...

Nice catch @jsha :+1:  ! I missed that part of the RFC, but now it does make sense !

In your description of `.send()`, what should we do if Transfer-Encoding is set but does not contain `chunked` ? If we stick to the spec, we should either automatically add `chunked` or error.

**Comment by algesten** [comment]
> In your description of .send(), what should we do if Transfer-Encoding is set but does not contain chunked ? If we stick to the spec, we should either automatically add chunked or error.

I think we can just error at this point – if someone ever complains about it we can fix later.

File: src/unit.rs [hunk 2]
```diff
@@ -47,12 +47,19 @@ impl Unit {
...
+                let is_transfer_encoding_set = !enc.is_empty();
+                let last_encoding = enc.split(',').last();
+                let is_chunked = last_encoding
+                    .map(|last_enc| last_enc.trim() == "chunked")
+                    .unwrap_or(false);
+                (is_transfer_encoding_set, is_chunked)
+            })
             // otherwise, no chunking.
-            .unwrap_or(false);
+            .unwrap_or((false, false));
```
Comment (algesten) [inline src/unit.rs:62]: I think this block could be more clear like this:

```rust
        let transfer_enc_header = req.header("transfer-encoding");

        let is_transfer_encoding_set =
            transfer_enc_header.is_some() && transfer_enc_header != Some("");

        let mut is_chunked =
            transfer_enc_header.and_then(|enc| enc.split(',').last()) == Some("chunked");
```

File: src/unit.rs [hunk 3]
```diff
@@ -64,8 +71,21 @@ impl Unit {
...
+                    BodySize::Known(size) => {
+                        extra.push(Header::new("Content-Length", &format!("{}", size)))
+                    }
+                    BodySize::Unknown => {
+                        if !is_transfer_encoding_set {
+                            extra.push(Header::new("Transfer-Encoding", "chunked"));
+                            is_chunked = true;
+                        }
+                    }
+                    BodySize::Empty => {}
```
Comment (algesten) [inline src/unit.rs:88]: This looks so nice now! ✨

**Comment by algesten** [comment]
Let's land this
